compiler/noirc_evaluator/src/ssa.rs

```diff
@@ -75,6 +75,31 @@ pub(crate) fn optimize_into_acir(
     time("SSA to ACIR", print_timings, || ssa.into_acir(&brillig, abi_distinctness))
 }
 
+/// Loop unrolling can return errors, since ACIR functions need to be fully unrolled.
+/// This meta-pass will keep trying to unroll loops and simplifying the SSA until no more errors are found.
+fn unroll_all_acir_loops(mut ssa: Ssa) -> Result<Ssa, RuntimeError> {
+    // Try to unroll loops first:
+    let mut unroll_errors;
+    (ssa, unroll_errors) = ssa.try_to_unroll_loops();
+
+    // Keep unrolling until no more errors are found
+    while !unroll_errors.is_empty() {
+        let prev_unroll_err_count = unroll_errors.len();
+
+        // Simplify the SSA before retrying
+        ssa = ssa.simplify_cfg();
+        ssa = ssa.mem2reg();
+
+        // Unroll again
+        (ssa, unroll_errors) = ssa.try_to_unroll_loops();
+        // If we didn't manage to unroll any more loops, exit
+        if unroll_errors.len() == prev_unroll_err_count {
```

Comment:
```suggestion
        if unroll_errors.len() >= prev_unroll_err_count {
```
We want it to be strictly decreasing, otherwise there could be a bug where we generate more errors each time and recur forever